packages/task/src/browser/task-service.ts

```diff
@@ -727,34 +730,48 @@ export class TaskService implements TaskConfigurationClient {
-        });
-        if (matchedRunningTaskInfo) { // the task is active
-            const taskName = this.taskNameResolver.resolve(task);
-            const terminalId = matchedRunningTaskInfo.terminalId;
-            if (terminalId) {
-                const terminal = this.terminalService.getByTerminalId(terminalId);
-                if (terminal) {
-                    if (TaskOutputPresentation.shouldSetFocusToTerminal(task)) { // assign focus to the terminal if presentation.focus is true
-                        this.terminalService.open(terminal, { mode: 'activate' });
-                    } else if (TaskOutputPresentation.shouldAlwaysRevealTerminal(task)) { // show the terminal but not assign focus
-                        this.terminalService.open(terminal, { mode: 'reveal' });
+        try {
+            const runningTasksInfo: TaskInfo[] = await this.getRunningTasks();
+            // check if the task is active
+            const matchedRunningTaskInfo = runningTasksInfo.find(taskInfo => {
+                const taskConfig = taskInfo.config;
+                return this.taskDefinitionRegistry.compareTasks(taskConfig, task);
+            });
+            console.debug(`running task ${JSON.stringify(task)}, already running = ${!!matchedRunningTaskInfo}`);
+
+            if (matchedRunningTaskInfo) { // the task is active
+                this.taskStartingLock.release(lockToken);
+                const taskName = this.taskNameResolver.resolve(task);
+                const terminalId = matchedRunningTaskInfo.terminalId;
+                if (terminalId) {
+                    const terminal = this.terminalService.getByTerminalId(terminalId);
+                    if (terminal) {
+                        if (TaskOutputPresentation.shouldSetFocusToTerminal(task)) { // assign focus to the terminal if presentation.focus is true
+                            this.terminalService.open(terminal, { mode: 'activate' });
+                        } else if (TaskOutputPresentation.shouldAlwaysRevealTerminal(task)) { // show the terminal but not assign focus
+                            this.terminalService.open(terminal, { mode: 'reveal' });
+                        }
                     }
                 }
+                const selectedAction = await this.messageService.info(`The task '${taskName}' is already active`, 'Terminate Task', 'Restart Task');
+                if (selectedAction === 'Terminate Task') {
+                    await this.terminateTask(matchedRunningTaskInfo);
+                } else if (selectedAction === 'Restart Task') {
+                    return this.restartTask(matchedRunningTaskInfo, option);
+                }
+            } else { // run task as the task is not active
+                console.debug('task about to start');
+                const taskInfo = await this.doRunTask(task, option);
+                this.taskStartingLock.release(lockToken);
```

Comment:
I think you should be able to move `this.taskStartingLock.release(lockToken);` into a `finally` block instead of duplicating the call.

Reply:
I need an early release in the other branch, so I thought it cleaner to only do a fallback release when an exception occurs.

Reply:
Ah, yes I didn't see the release in the other branch. Makes sense 👍